src/session_protocol.cpp

```diff
@@ -0,0 +1,651 @@
+        WrapInWSMessage,       // Wrap in the protobuf websocket message after enveloping
+        KeysEncryptMessage,    // Encrypt with the closed group keys after ennveloping
+    };
+
+    struct EncodeContext {
+        Mode mode;
+        // Parameters for BuildMode => Envelope
+        bool before_envelope_encrypt_for_recipient_deterministic;
+
+        // Ciphertext storing the result of encrypt for recipient deterministic, if it was necessary
+        // to encrypt before enveloping.
+        std::vector<uint8_t> before_envelope_ciphertext;
+
+        // Payload to set the envelope source if necessary.
+        std::optional<std::span<const uint8_t>> envelope_src;
+
+        // Type of message to mark the enevelope as
+        std::optional<SessionProtos::Envelope_Type> envelope_type;
+        AfterEnvelope after_envelope;
+    };
+
+    // Figure out how to encrypt the message based on the destination and setup the encoding context
+    EncodeContext enc = {};
+    switch (dest_type) {
+        case DestinationType::ClosedGroup: {
+            bool has_03_prefix =
+                    dest_closed_group_pubkey[0] == static_cast<uint8_t>(SessionIDPrefix::group);
+            if (has_03_prefix) {
+                if (space == config::Namespace::GroupMessages) {
+                    if (!dest_closed_group_keys)
+                        throw std::runtime_error(
+                                "API misuse: Sending to a closed group into the group messages "
+                                "namespace requires the closed group keys to be set");
+                    enc.mode = Mode::Envelope;
+                    enc.before_envelope_encrypt_for_recipient_deterministic = false;
+                    enc.after_envelope = AfterEnvelope::KeysEncryptMessage;
+                    enc.envelope_type =
+                            SessionProtos::Envelope_Type::Envelope_Type_CLOSED_GROUP_MESSAGE;
+                } else if (space == config::Namespace::RevokedRetrievableGroupMessages) {
+                    enc.mode = Mode::Plaintext;
+                } else {
+                    // Config messages should be sent directly rather than via this method (just
+                    // return plaintext and no-op). See:
+                    // https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L494
+                    enc.mode = Mode::Plaintext;
+                }
+            } else {
+                // Legacy closed groups which have a 05 prefixed key
+                enc.mode = Mode::Envelope;
+                enc.before_envelope_encrypt_for_recipient_deterministic = true;
+                enc.envelope_type =
+                        SessionProtos::Envelope_Type::Envelope_Type_CLOSED_GROUP_MESSAGE;
+                enc.after_envelope = AfterEnvelope::WrapInWSMessage;
+                enc.envelope_src = dest_closed_group_pubkey;
```

Comment:
to me, this should throw

Reply:
I need  confirmation that this branch can't get hit: https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L425

Reply:
The branch you've linked is the encryption for updated groups, legacy groups would fall into this case: https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L449

The code can be hit by a modified client at the moment, but most of the legacy groups logic has been removed so while a `VisibleMessage` would probably work, anything needed to update a legacy group state would fail - on a non-modified client legacy groups are in a read-only state so there _shouldn't_ be a way to send a message (ie. throwing in this case should be fine)

Reply:
Alright, updated, throws if you try and hit that codepath.